File: filelist/src/main/scala/farjs/filelist/popups/ExitController.scala
package farjs.filelist.popups

import farjs.filelist.FileListUiData
import farjs.ui.popup._
import farjs.ui.theme.Theme
import scommons.nodejs._
import scommons.react._

import scala.scalajs.js

object ExitController extends FunctionComponent[FileListUiData] {

  private[popups] var messageBoxComp: UiComponent[MessageBoxProps] = MessageBox

  protected def render(compProps: Props): ReactElement = {
    val props = compProps.wrapped
    val theme = Theme.useTheme().popup

    if (props.showExitPopup) {
      <(messageBoxComp())(^.plain := MessageBoxProps(
        title = "Exit",
        message = "Do you really want to exit FAR.js?",
        actions = js.Array(
          MessageBoxAction.YES { () =>
            props.onClose()
            process.stdin.emit("keypress", js.undefined, js.Dynamic.literal(
              name = "e",
              ctrl = true,
              meta = false,
              shift = false
            ))
          },
          MessageBoxAction.NO { () =>
            props.onClose()
          }
        ),
        style = theme.regular
      ))()
    }
    else null
  }
}

File: filelist/popups/ExitController.mjs
/**
 * @typedef {import("../FileListUi.mjs").FileListUiData} FileListUiData
 */
import React from "react";
import MessageBox from "@farjs/ui/popup/MessageBox.mjs";
import MessageBoxAction from "@farjs/ui/popup/MessageBoxAction.mjs";
import Theme from "@farjs/ui/theme/Theme.mjs";

const h = React.createElement;

/**
 * @param {FileListUiData} props
 */
const ExitController = (props) => {
  const { messageBoxComp } = ExitController;

  const theme = Theme.useTheme().popup;

  return props.showExitPopup
    ? h(messageBoxComp, {
        title: "Exit",
        message: "Do you really want to exit FAR.js?",
        actions: [
          MessageBoxAction.YES(() => {
            props.onClose();
            process.stdin.emit("keypress", undefined, {
              name: "e",
              ctrl: true,
              meta: false,
              shift: false,
            });
          }),
          MessageBoxAction.NO(() => {
            props.onClose();
          }),
        ],
        style: theme.regular,
      })
    : null;
};

ExitController.displayName = "ExitController";
ExitController.messageBoxComp = MessageBox;

export default ExitController;
